### PR TITLE
Move parts of type mapping to Core to facilitate type conversions

### PR DESCRIPTION
**Pull request by ajcvickers**
Issue #242

Also, start depending on runtime always having a type mapping associated with the property. This involves building the history model with conventions and updating tests that build the model manually.

### CODE REVIEW

File: src/EFCore/Metadata/Internal/CoreAnnotationNames.cs [hunk 1]
```diff
@@ -38,5 +38,11 @@ public static class CoreAnnotationNames
         ///     directly from your code. This API may change or be removed in future releases.
         /// </summary>
         public const string PropertyAccessModeAnnotation = "PropertyAccessMode";
+
+        /// <summary>
+        ///     This API supports the Entity Framework Core infrastructure and is not intended to be used
+        ///     directly from your code. This API may change or be removed in future releases.
+        /// </summary>
+        public const string TypeMapping = "Relational:TypeMapping"; // "Relational:" prefix to prevent breaking changes from 2.0
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Internal/CoreAnnotationNames.cs]: File an issue to fix this in 3.0

File: src/EFCore.Relational/Internal/RelationalPropertyExtensions.cs [hunk 2]
```diff
@@ -27,5 +29,13 @@ public static string FormatColumns([NotNull] this IEnumerable<IProperty> propert
...
         public static string GetConfiguredColumnType([NotNull] this IProperty property)
             => (string)property[RelationalAnnotationNames.ColumnType];
+
+        /// <summary>
+        ///     This API supports the Entity Framework Core infrastructure and is not intended to be used
+        ///     directly from your code. This API may change or be removed in future releases.
+        /// </summary>
+        public static RelationalTypeMapping FindRelationalMapping(
+            [NotNull] this IProperty property) 
+            => (RelationalTypeMapping)property[CoreAnnotationNames.TypeMapping];
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Internal/RelationalPropertyExtensions.cs]: Should this be `as RelationalTypeMapping` ?

File: src/EFCore/Metadata/Internal/CoreAnnotationNames.cs (hunk 1 above)
Comment (ajcvickers) [reply src/EFCore/Metadata/Internal/CoreAnnotationNames.cs]: @9913

File: src/EFCore.Relational/Internal/RelationalPropertyExtensions.cs (hunk 2 above)
Comment (ajcvickers) [reply src/EFCore.Relational/Internal/RelationalPropertyExtensions.cs]: Maybe. If it's a relational model, then any type mapping it has should be a RelationalTypeMapping. Not sure yet what the best way to enforce this is.

Comment (AndriySvyryd) [reply src/EFCore.Relational/Internal/RelationalPropertyExtensions.cs]: Do we really need to enforce it? As long as we can get a RelationalTypeMapping it should be fine and the relational validation already covers this

**Pull request merged**